File: ruby/phone-number/phone_number.rb
class PhoneNumber
  def initialize(input)
    @input = input
  end

  def number
    @number ||= clean_number
  end

  def area_code
    number[0..2]
  end

  def to_s
    "(#{area_code}) #{number[3..5]}-#{number[6..-1]}"
  end

  private

  def clean_number
    (digits.length == 10 && no_chars?) ? digits : '0000000000'
  end

  def digits
    remove_1(all_digits)
  end

  def remove_1(number)
    number.length == 11 && number.start_with?('1') ? number[1..-1] : number
  end

  def all_digits
    @input.scan(/\d+/).join
  end

  def no_chars?
    @input.scan(/[a-z]+/).empty?
  end
end

File: javascript/phone-number/babel-output/phone-number.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var PhoneNumber = (function () {
  function PhoneNumber(input) {
    _classCallCheck(this, PhoneNumber);

    this.input = input;
  }

  _createClass(PhoneNumber, [{
    key: "number",
    value: function number() {
      var digits = this.input.split("").filter(function (c) {
        return (/[0-9]/.test(c)
        );
      }).join("");
      return this.validate(digits);
    }
  }, {
    key: "validate",
    value: function validate(digits) {
      if (digits.length == 10) {
        return digits;
      } else if (digits.length === 11 && digits.startsWith("1")) {
        return digits.slice(1);
      } else {
        return "0".repeat(10);
      }
    }
  }, {
    key: "areaCode",
    value: function areaCode() {
      return this.number().slice(0, 3);
    }
  }, {
    key: "toString",
    value: function toString() {
      var number = this.number();
      var hiphenated = [number.slice(3, 6), number.slice(6)];
      return "(" + this.areaCode() + ") " + hiphenated[0] + "-" + hiphenated[1];
    }
  }]);

  return PhoneNumber;
})();

exports["default"] = PhoneNumber;
module.exports = exports["default"];